drivers/usb/udc/udc_smartbond.c

```diff
@@ -0,0 +1,1754 @@
+		if (GET_BIT(rxs0, USB_USB_RXS0_REG_USB_TOGGLE_RX0) != ep0_out_config->stat.data1) {
+			/* Toggle bit does not match discard packet */
+			REG_SET_BIT(USB_RXC0_REG, USB_FLUSH);
+			ep0_out_state->last_packet_size = 0;
+			LOG_WRN("Packet with incorrect data1 bit rejected");
+		} else {
+			read_rx_fifo(ep0_out_state,
+				     net_buf_tail(buf) + ep0_out_state->last_packet_size,
+				     fifo_bytes);
+			if (rxs0 & USB_USB_RXS0_REG_USB_RX_LAST_Msk) {
+				ep0_out_config->stat.data1 ^= 1;
+				net_buf_add(ep0_out_state->buf, ep0_out_state->last_packet_size);
+				if (ep0_out_state->last_packet_size < EP0_FIFO_SIZE ||
+				    ep0_out_state->buf->len == 0) {
+					k_work_submit_to_queue(udc_get_work_q(),
+							       &data->ep0_rx_work);
+				} else {
+					start_rx_packet(data, ep0_out_state);
+				}
+			}
+		}
+	}
+}
+
+static int udc_smartbond_ep_tx(const struct device *dev, uint8_t ep)
+{
+	struct usb_smartbond_data *data = dev->data;
+	struct smartbond_ep_state *ep_state = usb_dc_get_ep_in_state(data, USB_EP_GET_IDX(ep));
+	struct net_buf *buf;
+
+	if (udc_ep_is_busy(dev, ep) ||
+	    (ep_state->regs->epc_in & USB_USB_EPC1_REG_USB_STALL_Msk) != 0) {
+		return 0;
+	}
+
+	buf = udc_buf_peek(dev, ep);
+	LOG_DBG("TX ep 0x%02x len %u", ep, buf ? buf->len : -1);
+
+	if (buf) {
+		ep_state->buf = buf;
+		ep_state->last_packet_size = 0;
+
+		start_tx_packet(data, ep_state);
+
+		udc_ep_set_busy(dev, ep, true);
+	}
+
+	return 0;
+}
+
+static int udc_smartbond_ep_rx(const struct device *dev, uint8_t ep)
+{
+	struct usb_smartbond_data *data = dev->data;
+	struct smartbond_ep_state *ep_state = usb_dc_get_ep_out_state(data, USB_EP_GET_IDX(ep));
+	struct net_buf *buf;
+
+	if (udc_ep_is_busy(dev, ep)) {
+		return 0;
+	}
+
+	buf = udc_buf_peek(dev, ep);
+
+	if (buf) {
+		LOG_DBG("RX ep 0x%02x len %u", ep, buf->size);
+
+		ep_state->last_packet_size = 0;
+		ep_state->buf = buf;
+
+		start_rx_packet(data, ep_state);
+
+		udc_ep_set_busy(dev, ep, true);
+	}
+
+	return 0;
+}
+
+static int udc_smartbond_ep_enqueue(const struct device *dev, struct udc_ep_config *const ep_cfg,
+				    struct net_buf *buf)
+{
+	unsigned int lock_key;
+	const uint8_t ep = ep_cfg->addr;
+	int ret;
+
+	LOG_DBG("ep 0x%02x enqueue %p", ep, buf);
+	udc_buf_put(ep_cfg, buf);
+
+	if (ep_cfg->stat.halted) {
+		/*
+		 * It is fine to enqueue a transfer for a halted endpoint,
+		 * you need to make sure that transfers are re-triggered when
+		 * the halt is cleared.
+		 */
+		LOG_DBG("ep 0x%02x halted", ep);
+		return 0;
+	}
+
+	lock_key = irq_lock();
+
+	if (USB_EP_DIR_IS_IN(ep)) {
+		ret = udc_smartbond_ep_tx(dev, ep);
+	} else {
+		ret = udc_smartbond_ep_rx(dev, ep);
+	}
+
+	irq_unlock(lock_key);
+
+	return ret;
+}
+
+static int udc_smartbond_ep_dequeue(const struct device *dev, struct udc_ep_config *const ep_cfg)
+{
+	const uint8_t ep = ep_cfg->addr;
+	unsigned int lock_key;
+	struct net_buf *buf;
+
+	LOG_INF("ep 0x%02x dequeue all", ep);
+
+	lock_key = irq_lock();
+
```

Comment:
That looks too easy, do you need to find and cancel DMA transfers in progress?

Reply:
Your are right, can you give me a quick tip on how to trigger this case so I can verify that it works correctly?

Reply:
Fixed (scenario when this happens not tested though).

Reply:
There are few, (There was a simple one using shell commands to enqueue transfers, I have to bring it back). You can try CDC ACM sample, build it with CONFIG_SHELL and CONFIG_USBD_SHELL enabled. On Zephyr side enter shell command `usbd select sample_usbd`.
```

python3
>>> import usb
>>> dev = usb.core.find(idVendor=0x2fe3, idProduct=0x0001)
>>> dev.detach_kernel_driver(0)
>>> #to set DTR
>>> dev.ctrl_transfer(0x21, 0x22, 3, 0, 0)
0
>>> #Read all DCD and DSR notifications
>>> dev.read(0x81, 100)
array('B', [161, 32, 0, 0, 0, 0, 2, 0, 1, 0])
>>> dev.read(0x81, 100)
array('B', [161, 32, 0, 0, 0, 0, 2, 0, 3, 0])
>>> #On Zephyr side 'usbd disable' should trigger dequeue transfers for endpoint 0x01
```